### PR TITLE
Provide a generic interface to serialize and deserialize presenters.

### PR DESCRIPTION
**Pull request by vRallev**
This is useful to recreate presenters after a process has died.

### CODE REVIEW

**Comment by vRallev** [comment]
I know that you haven't decided what to with #39, yet. But this would be a good compromise in my opinion. If we want to use your library in our app, then we need a solution for this issue.

File: thirtyinch/src/main/java/net/grandcentrix/thirtyinch/internal/PresenterSavior.java [hunk 1]
```diff
@@ -52,7 +52,7 @@ public TiPresenter recover(final String id) {
     }
 
     public String safe(@NonNull final TiPresenter presenter) {
-        final String id = generateId(presenter);
+        final String id = presenter.getId();
         TiLog.v(TAG, "safe presenter with id " + id + " " + presenter);
```
Comment (passsy) [inline thirtyinch/src/main/java/net/grandcentrix/thirtyinch/internal/PresenterSavior.java]: Why is it better to have the presenter generate the id?

Comment (vRallev) [inline thirtyinch/src/main/java/net/grandcentrix/thirtyinch/internal/PresenterSavior.java]: Because it's the presenter's ID. Otherwise it would be an ID for a presenter used by the `PresenterSavior`. I'm using the ID at other places.

File: thirtyinch/src/main/java/net/grandcentrix/thirtyinch/TiConfiguration.java [hunk 2]
```diff
@@ -198,6 +199,16 @@ public Builder setUseStaticSaviorToRetain(final boolean enabled) {
             return this;
         }
 
+        /**
+         * An {@link TiPresenterSerializer} instance should be provided, if the {@link TiPresenter} should
+         * be restored, after a process has died, but the UI component is being recreated. Since this is
+         * a rare use case the default value is {@code null} and presenters aren't restored.
+         * @param serializer An implementation which can serialize and deserialize any presenter.
+         */
+        public Builder setPresenterSerializer(TiPresenterSerializer serializer) {
```
Comment (passsy) [inline thirtyinch/src/main/java/net/grandcentrix/thirtyinch/TiConfiguration.java]: Having the serializer as part of the configuration is interesting, I'll have to think about it

Comment (vRallev) [inline thirtyinch/src/main/java/net/grandcentrix/thirtyinch/TiConfiguration.java]: That was the ideal place for me. It's pretty handy with the global config and optional for tests.

File: thirtyinch/src/main/java/net/grandcentrix/thirtyinch/serialize/TiPresenterSerializer.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import net.grandcentrix.thirtyinch.TiPresenter;
+import net.grandcentrix.thirtyinch.TiView;
+
+/**
+ * Interface to serialize and deserialize presenters. The interface should be implemented, if
+ * a {@link TiPresenter} should be restored, after a process has died, but the UI component is
+ * being recreated.
+ */
+public interface TiPresenterSerializer {
```
Comment (vRallev) [inline thirtyinch/src/main/java/net/grandcentrix/thirtyinch/serialize/TiPresenterSerializer.java]: @passsy After reviewing all the changes in the 0.8.0 version I think this interface can be removed and the new `TiPresenterSavior` could be used instead. After that it would be possible for me to release the module in #51 separately and I could finally stop working on the fork.

The only changes that would be necessary to make this work is to add the `TiPresenterSavior` as argument in the `TiConfiguration`. This way I get a chance to replace the default savior. 

The 2nd change that's necessary is to modify the recover method in the savior interface `TiPresenter recover(String presenterId, @NonNull Object host);`. Here a third parameter would be necessary that you've already implemented: `TiPresenterProvider`. In most cases that would be the `TiFragment` or `TiActivity`. That would give me the option to construct a new presenter in case it can be recovered after the process died.

Would that be possible?

Comment (vRallev) [inline thirtyinch/src/main/java/net/grandcentrix/thirtyinch/serialize/TiPresenterSerializer.java]: I think both changes are safe, because nothing of that was exposed to the public API, yet. Do you want me to create a different PR?

Comment (vRallev) [inline thirtyinch/src/main/java/net/grandcentrix/thirtyinch/serialize/TiPresenterSerializer.java]: Dang, I just realized that it's not necessary to make any changes in the config class, because in order to access the config, you need to create a presenter first. 

There's no way to replace the savior in the delegate from `TiFragment` or `TiActivity` at the moment. I think it's better to add a new provider interface for the savior and the default implementation delegate to the `PresenterSavior`. Although a global config would be nicer.

Comment (passsy) [inline thirtyinch/src/main/java/net/grandcentrix/thirtyinch/serialize/TiPresenterSerializer.java]: I'll look into presenter serialisation for the next release and we'll find a way to provide you an API.
The `PresenterSavior` is the new heart of this library. I don't think the savior will be replaceable with a third party impl because small mistakes could lead to major leaks. Instead I'm thinking about APIs for the savior to control the presenter store. Haven't thought this through completely but you get the direction.

**Comment by passsy** [comment]
I still don't get why a new API is required. I implemented saving presenter state with `android-state` in a few seconds. What is missing?

see https://github.com/grandcentrix/ThirtyInch/compare/feature/save_state

This implementation could be moved into a `BasePresenter`/`BaseActivity` in your project. A fork is not required.

**Comment by vRallev** [comment]
This would be tied to the activity lifecycle again. That's not what you want in case a longer operation is going on. Then there wouldn't be a good hook to restore the presenter.
